docs/manual.md

```diff
@@ -807,91 +805,294 @@ user:
+```
+
+When a transformed test fails, it will detail the transformers used, the `-o include_raw` option can be used to include the raw, untransformed attribute value:
+```
+$ goss v
+F
+
+Failures/Skipped:
+
+KernelParam: net.core.somaxconn: value:
+Expected
+    128
+to be >
+    200
+the transform chain was
+    [{"to-numeric":{}}]
+
+Total Duration: 0.001s
+Count: 1, Failed: 1, Skipped: 0
+
+
+$ goss v -o include_raw
+F
 
-$ goss validate
-..
+Failures/Skipped:
+
+KernelParam: net.core.somaxconn: value:
+Expected
+    128
+to be >
+    200
+the transform chain was
+    [{"to-numeric":{}}]
+the raw value was
+    "128"
 
 Total Duration: 0.001s
-Count: 2, Failed: 0
+Count: 1, Failed: 1, Skipped: 0
+
 ```
 
-## Advanced Matchers
-Goss supports advanced matchers by converting json input to [gomega](https://onsi.github.io/gomega/) matchers.
+### Advanced Matchers
 
-### Examples
+Goss supports advanced matchers by converting YAML input to [gomega](https://onsi.github.io/gomega/) matchers.
+
+#### String Matchers
 
-Validate that user `nobody` has a `uid` that is less than `500` and that they are **only** a member of the `nobody` group.
+These will convert the system attribute to a string prior to matching.
 
+* `'55'` - Checks that the numeric is "55" when converted to string
+* `have-prefix: pre` - Checks if string starts with "pre"
+* `have-suffix: suf` - Checks if string ends with "suf"
+* `match-regexp: '.*'` - Checks if string matches regexp
+* `contain-substring: '2'` - Checks if string contains "2"
+
+Example:
 ```yaml
-user:
-  nobody:
-    exists: true
-    uid:
-      lt: 500
-    groups:
-      consist-of: [nobody]
+matching:
+  example:
+    content: 42
+    matches:
+      and:
+        - '42'
+        - have-prefix: '4'
+        - have-suffix: '2'
+        - match-regexp: '\d{2}'
+        - contain-substring: '2'
 ```
 
-Matchers can be nested for more complex logic, for example you can ensure that you have 3 kernel versions installed and none of them are `4.1.0`:
+#### Numeric matchers
 
+These will convert the system attribute to a numeric prior to matching.
+
+* `42` - If the expected type is a number
+* `gt, ge, lt, le` - Greater than, greater than or equal, less than, etc..
+
+Example:
 ```yaml
-package:
-  kernel:
-    installed: true
-    versions:
+matching:
+  example:
+    content: "42"
+    matches:
+      and:
+        - 42
+        - 42.0
+        - gt: 40
+        - lt: 45
+```
+
+#### Array matchers
+
+These will convert the system attribute to an array prior to matching. Strings are split on "\n"
+
+
+* `contain-element: matcher` - Checks if the array contains an element that passes the matcher
+* `contain-elements: [matcher, ...]` - checks if the array is a superset of the provided matchers
+* `[matcher, ...]` - same as above
+* `equal: [value, ...]` - Checks if the array is exactly equal to provided array
+* `consist-of: [matcher, ...]` - Checks if the array consists of the provided matchers (order does not matter)
```

Comment:
I haven't, from this and the example below, figured out where I would use `contain-element` vs `consist-of`. I find it difficult to understand partly because both of them allow matchers to be present.

I can think of wanting to:

* assert that an array contains particular elements/matches (without caring about ordering)
* assert that an array contains a sequence of particular elements/matches (e.g. contains a then any-single-element then b, but that sequence can be anywhere in the array)
  * same but a then not-b then b
* assert that an array contents matches sequence x, y, z exactly

I think I'm struggling partly because the doc prose is more of the reference style (which I find quite abstract, though of course that's highly subjective!) rather than how-to/use-case style. I think if it contained more worked examples I'd have an easier time.

Reply:
`contain-element` is there since 2016. This PR adds `contain-elements` which is really just a subset check.

For your scenarios:
> assert that an array contains particular elements/matches (without caring about ordering)

contain-element / contain-elements / ["a", "b"] (this is an alias for contain-elements) will satisfy this

> assert that an array contents matches sequence x, y, z exactly

equal

> assert that an array contains a sequence of particular elements/matches (e.g. contains a then any-single-element then b, but that sequence can be anywhere in the array)
same but a then not-b then b

I don't think this is supported today, sounds like a subset of matchers where order matters. I'm curious about the real-world use-case of this.. can be a new matcher that's added in 4.X

> I think if it contained more worked examples I'd have an easier time.

I love this suggestion, I wonder if it should be:
1. A new section in manual.md
2. A new doc that's linked to from manual.md (examples.md?)
3. Just an example folder(s) of yaml files. Could use either the `matching` or maybe real world examples?

Thoughts?

Reply:
I think a new section of the manual.md would work well.

Would you be interested in a PR that adds https://www.mkdocs.org/ -> github pages for documentation?